stl/inc/span

```diff
@@ -0,0 +1,665 @@
+        return _Extent;
+    }
+};
+
+template <>
+struct _Span_extent_type<dynamic_extent> {
+    constexpr _Span_extent_type() noexcept = default;
+
+    constexpr explicit _Span_extent_type(const size_t _Size) noexcept : _Mysize(_Size) {}
+
+    template <size_t _OtherExtent>
+    constexpr explicit _Span_extent_type(const _Span_extent_type<_OtherExtent> _Ext) noexcept : _Mysize(_Ext._Size()) {}
+
+    _NODISCARD constexpr size_t size() const noexcept {
+        return _Mysize;
+    }
+
+private:
+    size_t _Mysize{0};
+};
+
+// VARIABLE TEMPLATE _Is_span_v
+template <class _Ty, size_t _Extent>
+class span;
+
+template <class>
+inline constexpr bool _Is_span_v = false;
+
+template <class _Ty, size_t _Extent>
+inline constexpr bool _Is_span_v<span<_Ty, _Extent>> = true;
+
+// VARIABLE TEMPLATE _Is_std_array_v
+template <class>
+inline constexpr bool _Is_std_array_v = false;
+
+template <class _Ty, size_t _Extent>
+inline constexpr bool _Is_std_array_v<array<_Ty, _Extent>> = true;
+
+// VARIABLE TEMPLATE _Is_span_convertible_range_v
+template <class _Cont, class _Ty>
+inline constexpr bool _Is_span_convertible_range_v =
+    is_convertible_v<remove_pointer_t<decltype(_STD data(_STD declval<_Cont&>()))> (*)[], _Ty (*)[]>;
+
+// VARIABLE TEMPLATE _Has_container_interface_v
+template <class, class = void>
+inline constexpr bool _Has_container_interface_v = false;
+
+template <class _Cont>
+inline constexpr bool _Has_container_interface_v<_Cont,
+    void_t<decltype(_STD data(_STD declval<_Cont&>())), decltype(_STD size(_STD declval<_Cont&>()))>> = true;
+
+// VARIABLE TEMPLATE _Is_span_compatible_container_v
+// clang-format off
+template <class _Cont, class _Ty>
+inline constexpr bool _Is_span_compatible_container_v =
+       !_Is_span_v<remove_const_t<_Cont>>
+    && !_Is_std_array_v<remove_const_t<_Cont>>
+    && !is_array_v<_Cont>
+    && _Is_span_convertible_range_v<_Cont, _Ty>
+    && _Has_container_interface_v<_Cont>;
+// clang-format on
+
+// [views.span]
+// CLASS TEMPLATE span
+template <class _Ty, size_t _Extent = dynamic_extent>
+class span : public _Span_extent_type<_Extent> {
+public:
+    using _Mybase                = _Span_extent_type<_Extent>;
+    using element_type           = _Ty;
+    using value_type             = remove_cv_t<_Ty>;
+    using index_type             = size_t;
+    using difference_type        = ptrdiff_t;
+    using pointer                = _Ty*;
+    using const_pointer          = const _Ty*;
+    using reference              = _Ty&;
+    using const_reference        = const _Ty&;
+    using iterator               = _Span_iterator<_Ty>;
+    using const_iterator         = _Span_iterator<const _Ty>;
+    using reverse_iterator       = _STD reverse_iterator<iterator>;
+    using const_reverse_iterator = _STD reverse_iterator<const_iterator>;
+
+    static constexpr index_type extent = _Extent;
+
+    // [span.cons] Constructors, copy, and assignment
+    constexpr span() noexcept = default;

```

Comment:
Missing constraint: https://eel.is/c++draft/span.cons#1

Reply:
Hm I am getting a lot of errors with
```cpp
	template <enable_if_t<_Extent == dynamic_extent || _Extent == 0, int> = 0 >
	constexpr span() noexcept {};
```
Error	C2938	'std::enable_if_t<false,int>' : Failed to specialize alias template	SpanTest	

Maybe there are some incorrect tests and the error message is just not helpfull

Reply:
I have added a second specialization `_Span_extent<0>` and deleted the default constructor for all other static spans.

Reply:
This happens when it's possible for your constraint to be eagerly evaluated; i.e. because it depends on `_Extent` which is a class template parameter, the compiler can attempt to instantiate `enable_if_t` when the class is mentioned, instead of when the constructor is called. We face this issue regularly in classes like `pair` and `tuple`. The solution is to add an extra layer of indirection: `template <size_t _Extent2 = _Extent, enable_if_t<CONDITION USING _Extent2`. Here, the compiler doesn't know what `_Extent2` will be (even though the default value `_Extent` will always be used), so it won't evaluate the constraint until the constructor is actually needed.

Reply:
Thanks a lot for the explanation. 

Its amazing how much there is still to learn, but also quite terrifying what I learn.